R/death_functions.R

```diff
@@ -892,28 +910,34 @@ death_injury_matrix_count <- function(ph.data,
           x_combo <- x_combo[, list(deaths = sum(deaths)), by = setdiff(names(x_combo), c("deaths"))]
         } else {
           x_combo[, intent := 'Any intent']
-          x_combo <- x_combo[, list(deaths = sum(deaths), temp_ypll = sum(get(paste0("ypll_", ypll_age)))),
-                             by = setdiff(names(x_combo), c("deaths", paste0("ypll_", ypll_age)))]
+          x_combo <- x_combo[, list(deaths = sum(deaths), temp_ypll = sum(get(ypll_col_name))),
+                             by = setdiff(names(x_combo), c("deaths", ypll_col_name))]
           setnames(x_combo, "temp_ypll", paste0("ypll_", ypll_age))
         }
       }
 
       if(mechanism == 'none'){x_combo[, mechanism := 'Any mechanism']}
 
     # Create rows for zero values (otherwise rows would simply be missing) ----
-      # create temporary vectors of unique values of all columns EXCEPT deaths and ypll_##
-        for(i in setdiff(names(x_combo), c('deaths', grep('^ypll_', names(x_combo), value = T)))){
-          assign(paste0('xyz_', i), unique(x_combo[, get(i)]))
-        }
+      # Select columns to use for combination
+        cols_to_use <- setdiff(names(x_combo), c('deaths', grep('^ypll_', names(x_combo), value = TRUE)))
 
-        if(myorig.intent == '*'){xyz_intent <- unique(death_injury_matrix()[]$intent)}
+      # Create list of unique values for each selected column
+        unique_col_vals <- lapply(cols_to_use, function(col) unique(x_combo[[col]]))
 
-      # create template of all combinations of x_combo values
-        template.xyz <- setDT(expand.grid(mget(ls(pattern = 'xyz_'))))
-        setnames(template.xyz, gsub('^xyz_', '', names(template.xyz)))
+      # Create names for the list
+        names(unique_col_vals) <- c(cols_to_use)
 
-      # merge actual values onto template.xyz
-        x_combo <- merge(template.xyz, x_combo, all = T)
+      # If myorig.intent is '*', add intent column
+        if (exists("myorig.intent") && myorig.intent == '*') {
+          unique_col_vals$intent <- unique(death_injury_matrix()$intent)
+        }
+
+      # Use CJ to create all combinations
+        template_xyz <- do.call(CJ, unique_col_vals)
+
+      # Merge actual values onto template_xyz
+        x_combo <- merge(template_xyz, x_combo, all = T)
```

Comment:
This should be all.X =T, not all =T, yeah? Otherwise what is the point of template_xyz?

Reply:
Addressed with commit fc1a9d8f66d1264416d04378f22683a78d37c9e6

- I appreciate the attention to detail! 
- After review, I think it is correct as is because `template_xyz` is created to expand `x_combo`. For example, if we are grouping deaths by race/eth and gender and there are no Males who are Multiple Race, `template_xyz` will provide this combination and the number of deaths in that combination will be set to zero in the following line (`x_combo[is.na(deaths), deaths := 0]`). 
- I added comments to the all.x and all.y arguments to make the logic more explicit.
- Fixed in `death_xxx_count`, `death_injury_matrix_count`, and `death_other_count`.